src/main/kotlin/uk/gov/justice/digital/hmpps/hmppspersononprobationuserapi/service/DelegatedAccessService.kt

```diff
@@ -0,0 +1,126 @@
+      )
+    }
+    if (!delegatedAccessPermissionList.isNullOrEmpty()) {
+      throw ValidationException("Revoke all permissions granted, before removing the access id $id")
+    }
+    delegatedAccessEntity.deletedDate = now
+    return delegatedAccessRepository.save(delegatedAccessEntity)
+  }
+
+  @Transactional
+  fun getAllAccessByInitiatorUserId(id: Int): List<DelegatedAccessEntity> {
+    val accessList = delegatedAccessRepository.findByInitiatedUserId(id.toLong())
+    return accessList
+  }
+
+  @Transactional
+  fun getActiveAccessByInitiatorUserId(id: Int): List<DelegatedAccessEntity> {
+    val accessList = delegatedAccessRepository.findByInitiatedUserIdAndDeletedDateIsNull(id.toLong())
+    return accessList
+  }
+
+  @Transactional
+  fun grantDelegatedAccessPermission(accessId: Int, permissionId: Int): DelegatedAccessPermissionEntity {
+    val now = LocalDateTime.now()
+    val delegatedAccess = delegatedAccessRepository.findById(accessId.toLong()) ?: throw ResourceNotFoundException("User with id $accessId not found in database ")
+    val delegatedAccessPermissionAlreadyExists = delegatedAccessPermissionRepository.findByDelegatedAccessIdAndPermissionIdAndGrantedIsNotNullAndRevokedIsNull(accessId.toLong(), permissionId.toLong())
+    if (delegatedAccessPermissionAlreadyExists != null) {
+      throw DuplicateDataFoundException("Permission already granted and active!")
+    }
+    val delegatedAccessPermissionEntity = DelegatedAccessPermissionEntity(
+      id = null,
+      delegatedAccessId = accessId,
+      permissionId = permissionId,
+      granted = now,
+      revoked = null,
+    )
+    return delegatedAccessPermissionRepository.save(delegatedAccessPermissionEntity)
+  }
+
+  @Transactional
+  fun revokeDelegatedAccessPermission(accessId: Int, permissionId: Int): DelegatedAccessPermissionEntity {
+    val now = LocalDateTime.now()
+    val delegatedAccessPermissionEntity = delegatedAccessPermissionRepository.findByDelegatedAccessIdAndPermissionIdAndGrantedIsNotNullAndRevokedIsNull(accessId.toLong(), permissionId.toLong()) ?: throw ResourceNotFoundException("Given access id $accessId and permission id $permissionId not found in the database or permission already revoked!")
+    delegatedAccessPermissionEntity.revoked = now
+    return delegatedAccessPermissionRepository.save(delegatedAccessPermissionEntity)
+  }
+
+  @Transactional
+  fun getAllAccessPermissionByUserId(userId: Int): MutableList<DelegatedAccessPermissionEntity?> {
+    val accessPermissionList = mutableListOf<DelegatedAccessPermissionEntity?>()
+    getAllAccessByInitiatorUserId(userId).forEach {
+      val list = it.id?.let { it1 -> delegatedAccessPermissionRepository.findByDelegatedAccessId(it1) }
+      if (list != null) {
+        accessPermissionList.addAll(list)
+      }
+    }
+    return accessPermissionList
+  }
+
+  @Transactional
+  fun getActiveAccessPermissionByInitiatorUserId(userId: Int): MutableList<DelegatedAccessPermissionEntity?> {
```

Comment:
the function says find active but you don't seem to be checking whether they are revoked? What's this supposed to mean

Reply:
Fix provided!